### ISSUE TITLE
inference result error

### ISSUE THREAD

**Issue by Tomhardy13** [TexasInstruments/edgeai-yolov5#16]
## ❔Question
hi,thanks for you release the code,I run detect.py test my data,I find the keypoint of the distant person's head is far from the detection,I use Yolov5s6_pose_960.pt input size set 960,1080,1280.

## Additional context
like this ,input size set 960
![posetest](https://user-images.githubusercontent.com/41933588/172291667-5ca3bc6c-334d-4aaa-aecf-6b7c969beca2.png)

when I set 1080, more keypoint of head far from the detection
![posetest_1080](https://user-images.githubusercontent.com/41933588/172292285-19374f57-3775-4114-b81e-83acfb1f8459.png)

**Comment by debapriyamaji** [comment]
@Tomhardy13 , Thanks for trying it out, I haven't tested the detect.py script. I will look into it. It's probably some issue in visualization.  

Regards, Debapriya

**Comment by wangduyang** [comment]
你应该使用的是预训练模型进行推理了。
换一下模型吧

**Comment by xiakj** [comment]
watching

**Comment by isagastiberri** [comment]
Hello! I'm also using detect.py and having the same issue. I'm wondering, should we use another script for inference then?

**Comment by tucachmo2202** [comment]
> 

Hi, @debapriyamaji. I have run the detect code and met same problem. I see that the wrong keypoint sometime have pretty high confidence so I have to check keypoint outside bounding box.

**Comment by Tomhardy13** [comment]
Like @isagastiberri ,I also found the keypoint of the left eye and left ear wrong. Consider that weights with input sizes of 640 and 960 both cause this error,so It is possible that the data is causing this error. @debapriyamaji  Have you processed the data set?

**Comment by debapriyamaji** [comment]
Hi @Tomhardy13 , Can you please share the sample image with which you are seeing this problem.

**Comment by isagastiberri** [comment]
I'm now retraining the network in case there is any kind of mistake in the weights. In the training batches that are generated during training there aren't any mistakes. During inference, the error only happens with left ear and eye, and it only fails sometimes. So it doesn't seem to be a visualization error. @debapriyamaji I'll share a couple of my own images here too, in case it's useful:
![20211006_210321_5012JRM_DESKTOP-OLN8HBB_frame_170_sensor_1](https://user-images.githubusercontent.com/56122555/176175712-49f23a22-a327-4ed0-9864-94fe2d457ab8.jpg)

![20211006_210321_5012JRM_DESKTOP-OLN8HBB_frame_170_sensor_1](https://user-images.githubusercontent.com/56122555/176175845-20535e5f-0120-4952-9ef9-2fa6b859681f.jpg)
I've tried with fisheye and rectified images and the error still appears. Anyways, @Tomhardy13 has the same error as myself and his images don't have any distortion, so that shouldn't be the problem in this case. If I manage to narrow it down more I'll update you guys. And thanks for looking into it!

**Comment by debapriyamaji** [comment]
@isagastiberri , Can you please share some input rectified images so that I can test them on the network. I couldn't reproduce this on the COCO dataset. 

Regards, Debapriya

**Comment by Caius-Lu** [comment]
> @isagastiberri , Can you please share some input rectified images so that I can test them on the network. I couldn't reproduce this on the COCO dataset.
> 
> Regards, Debapriya

I also get this error in coco datasets and my own datasets.
Coco datasets:
0000000026690.jpg
0000000036494.jpg
...
The head of keypoints error like @isagastiberri .
In my own datasets, i find that many kepoints coordinate value  is negative. 
some kepoints coordinate value is 88 before scale_coords, and turn 0.
 
Regards

**Comment by Tomhardy13** [comment]
This zip file contains some images,the data has errors in the keypoint of the left ear and left eye.



------------------ 原始邮件 ------------------
发件人:                                                                                                                        "TexasInstruments/edgeai-yolov5"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年6月28日(星期二) 晚上7:21
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [TexasInstruments/edgeai-yolov5] inference result error (Issue #16)





 
Hi @Tomhardy13 , Can you please share the sample image with which you are seeing this problem.

**Comment by Caius-Lu** [comment]
Despite the errors data,  the  keypoints which as the part of head  offset  to big.

在2022-06-29 ***@***.***写道：

This zip file contains some images,the data has errors in the keypoint of the left ear and left eye.



------------------ 原始邮件 ------------------
发件人: "TexasInstruments/edgeai-yolov5" ***@***.***&gt;;
发送时间:&nbsp;2022年6月28日(星期二) 晚上7:21
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [TexasInstruments/edgeai-yolov5] inference result error (Issue #16)






Hi @Tomhardy13 , Can you please share the sample image with which you are seeing this problem.

**Comment by isagastiberri** [comment]
[test-data-isagastiberri.zip](https://github.com/TexasInstruments/edgeai-yolov5/files/9008580/test-data-isagastiberri.zip)
@debapriyamaji I uploaded 30 images to test. How are you extracting the inference in your images? I use the detect.py script like

 `python detect.py --weights yolov5m6_pose_960.pt --source ../face1_fisheye/ --img-size 960 960 --conf-thres 0.3 --kpt-label`

**Comment by isagastiberri** [comment]
@debapriyamaji I've retrained the network by running

`python train.py --data data/coco_kpts.yaml --cfg models/hub/yolov5m6_kpts.yaml --weights yolov5m6_person_960.pt --batch-size 16 --img 960 --kpt-label --hyp yolov5m6_pose_960_hyp.yaml --epochs 100 --cache-images`

I found that, even in the training, the predictions show the leftside keypoints outside the bounding boxes. Checking the labels, those are fine, so it doesn't seem to be a visualization mistake, but I think it's weird that it has such an specific error. Did you neves find this error? Is it possible that you uploaded a different version than the one you were locally using or maybe you changed something afterwards? These are the examples I mean, generated during training.

Labels:
![image](https://user-images.githubusercontent.com/56122555/177111376-a3c0882e-b672-4982-9fc3-544245abef5b.png)

Predictions:
![Uploading image.png…]()

**Comment by BaofengZan** [comment]
I have the same issue.    Is there any way to solve？

**Comment by debapriyamaji** [comment]
@isagastiberri I was looking into the issue and I was able to reproduce it. I am not sure why I missed it during training.

I have one request: Will you please try inference on the *ti-lite models and let me know if you see any issue with them.
https://github.com/TexasInstruments/edgeai-yolov5/tree/yolo-pose#yolov5-ti-lite-based-models-and-ckpts

I have started re-training some of the models and they seem to be working fine. Once, you confirm that the *ti-lite models are working fine, I can share the newly trained models for verification. Thanks again for reporting the issue.

Regards, Debapriya

**Comment by debapriyamaji** [comment]
I have updated the models. Let me know if you see any issue with the latest models.
https://github.com/TexasInstruments/edgeai-yolov5/tree/yolo-pose#yolo-pose-models-and-ckpts

Regards, Debapriya

**Comment by isagastiberri** [comment]
@debapriyamaji sorry, I was out of office for a while and I'm just seeing this. I'll try the updated models to confirm. Regardless, thanks a lot for the work!

**Comment by isagastiberri** [comment]
I checked and everything is fine with the updated weights, thanks a lot!

**Comment by debapriyamaji** [comment]
@isagastiberri  Thanks a lot for confirming. There were some issues while training at your place as well. 

It will be great if you can confirm whether you are still seeing any issues in training.

Regards, Debapriya

**Comment by tctco** [comment]
Hi @debapriyamaji thanks for your amazing job!

I have tried to finetune the `*ti-lite models` on my custom dataset and the result looks okay. However, when I tried to use `yolov5-nano` model as the network backbone and retrained the model, the 'out-of-bbox' error still occurs (the left leg and left ear). I wonder if this is a bug in the training code. Is there a way to fix this?

![image](https://user-images.githubusercontent.com/45505657/188647151-90f55be5-8a5a-4830-b829-d0251e2cb1f1.png)
